### DISCUSSION TITLE
Absolute path is being read a relative

### DISCUSSION THREAD

**Discussion by akshaybabloo** [typst/typst#4195] [General]
### Description

I am trying to read a file that is not part of the folder I am running the typst from and when I am trying to read that file, it converts an d absolute path and attaches it to relative path

```typ
#let text = read("/home/akshay/code/personal/some/thing/temo2.py")
#raw(text, lang: "python")
```

And this is the error I get

```
typst compile temp.typ
error: file not found (searched at /home/akshay/code/personal/phd-thesis/home/akshay/code/personal/some/thing/temo2.py)
  ┌─ temp.typ:3:17
  │
3 │ #let text = read("/home/akshay/code/personal/head-tracker/non-nn/encoder.py")
  │                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

### Operating system

Linux

### Typst version

- [X] I am using the latest version of Typst

**Comment by Enivex** [comment]
Paths are resolved relative to the root folder, which by default is the folder your typ file is in. You cannot access files outside that root folder. You would therefore have to specify a common root folder when compiling.

**Reply by wrzian** [reply]
Maybe a docs update or a more informative error could be issued if you try to access a path that looks like `/abc/...` ? Feels like a hint warning could be a good solution?

**Comment by AhmedSalih3d** [comment]
But why is this not possible?

Suppose I have a template where I want to feed it a file holding my glossary values, why should it be complex/disallowed to do so?

I find typst to be very modern, but having some weird restrictions

**Reply by fabian-s** [reply]
Seconding this -- what a weird way to subvert user expectations that are honored by basically every other tool.

**Comment by dedebenui** [comment]
I have a Python app all neatly structured in a directory, which I can deploy on a remote server. I don't want this directory to contain actual data because it may be overridden by deploying a new version. Therefore, I have a separate data folder in the home directory. I intend to use a Typst template in my app directory and compile a new pdf with user data. Thanks to `#sys.input`, I don't have to do any templating shenanigans, I just call `compile` with a different `--input`. The problem is that my input contains paths to files located in the data folder, outside the app directory and thus, outside where `main.typ` is located.

Here is a schematic of the situation:
```
home_folder
├── app
│   └── src
│       ├── app.py
│       └── data
│           └── main.typ
└── data
    ├── final_document.pdf
    ├── generated_image.svg
    └── user_data.json
```

- `main.typ` is a data file bundled with the app
- `user_data.json` contains an absolute path to `generated_image.svg`
- `app.py` wants to call typist to `compile --input data=/home/user/data/user_data.json main.typ` into `generated_document.pdf` using data it generated earlier as `generated_image.svg`, referenced in  `user_data.json`.

Without support for absolute paths, I would need to do the following:
- generate a `user_data_typst.json` solely for the sake of Typst, to accommodate relative-paths-only
- programatically create a new folder, copy everything in it, compile, move the pdf out, and delete everything

That seems really convoluted (almost like using LaTeX). What's the intention behind not allowing absolute paths?

**Reply by Enivex** [reply]
@dedebenui

The typst compiler has a `--root` argument for situations like this. You just have to choose a root that incorporates everything you want to use

![image](https://github.com/user-attachments/assets/1b0565ee-99df-4849-8e1c-9757afc4caa9)

**Reply by dedebenui** [reply]
Thank you for the very quick reply. If I understand correctly, although it says `for absolute paths`, it means `from where do I resolve absolute paths?`, which is a stretched interpretation of 'absolute'... If I set it to the root of the file system, will relative paths still work relative to the current working directory? If I set `--root /user/home_folder`, then I still have to make sure every path in every situation is relative to home and not absolute. In all cases, it's all pretty unintuitive compared to the standard POSIX  system of absolute paths, relative paths and working directory. In the end I went for the `create a temporary folder with everything copied in it` approach.

**Reply by fabian-s** [reply]
Seconding this -- not a good look at all to subvert user expectations that are honored by basically every other tool in this way, and use misleading terminology on top of that.  
Is there any technical reason why you can't allow conventional syntax for absolute and relative paths?

**Comment by laurmaedje** [comment]
Reasoning is now explained in https://github.com/typst/typst/issues/5382#issuecomment-5645768707.
